### PR TITLE
Adding Security Features

### PR DESCRIPTION
**Pull request by elcore**
Creating a secure OpenSSH Connection by disabling SHA-1.

### CODE REVIEW

**Comment by moul** [comment]
Hi @elcore, thanks for contributing, I'm not aware of this kind of security improvements, I will review your commit

File: 15.04/patches/etc/ssh/sshd_config [hunk 1]
```diff
@@ -87,3 +87,9 @@ Subsystem sftp /usr/lib/openssh/sftp-server
 # and ChallengeResponseAuthentication to 'no'.
 UsePAM yes
 UseDNS no
+
+#### OpenSSH Security ####
+
+KexAlgorithms curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
```
Comment (moul) [inline 15.04/patches/etc/ssh/sshd_config]: Default value seems to be: **curve25519-sha256@libssh.org**,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,**diffie-hellman-group-exchange-sha256**,diffie-hellman-group14-sha1

File: 15.04/patches/etc/ssh/sshd_config [hunk 2]
```diff
@@ -87,3 +87,9 @@ Subsystem sftp /usr/lib/openssh/sftp-server
...
+Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
```
Comment (moul) [inline 15.04/patches/etc/ssh/sshd_config]: Default value is: **aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com,chacha20-poly1305@openssh.com** (the same but with a different order)

File: 15.04/patches/etc/ssh/sshd_config [hunk 3]
```diff
@@ -87,3 +87,9 @@ Subsystem sftp /usr/lib/openssh/sftp-server
...
+MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-ripemd160-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,hmac-ripemd160,umac-128@openssh.com
```
Comment (moul) [inline 15.04/patches/etc/ssh/sshd_config]: Default value is: **umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512** (different order)

**Comment by moul** [comment]
I think that changing the order is a good idea, but I'm not sure about removing some `Kex Algorithms`, I don't know if it may prevent some people to connect to their new server

@elcore, what do you think about adding older algorithms at the end of the value so we can privilege security for modern clients but keep access for older ones ?

These 3 options are available for both *sshd* and *ssh*(-client), so in addition we may suggest people to update their ~/.config file in our documentation

We need more opinions, cc @scaleway/core-team

**Comment by elcore** [comment]
Hi @moul,

changing the order is a really good idea :-). 

This would be perfect : "so in addition we may suggest people to update their ~/.config file in our documentation".

"@elcore, what do you think about adding older algorithms at the end of the value so we can privilege security for modern clients but keep access for older ones ?" :

I and @stribika have the same point of view (https://stribika.github.io/2015/01/04/secure-secure-shell.html), but this would be a compromise. I would prefer to ask our Community.

This is true : "We need more opinions, cc @scaleway/core-team"

**Comment by elcore** [comment]
If we add Support for older Clients we should change this

```
#### OpenSSH Security #### 

KexAlgorithms curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha1
Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-ripemd160-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,hmac-ripemd160,umac-128@openssh.com
```
If we just Support new Clients we should change this (removing unused HostKeys)

```
[...}
HostKey /etc/ssh/ssh_host_rsa_key
HostKey /etc/ssh/ssh_host_ed25519_key
[...}

#### OpenSSH Security #### 

KexAlgorithms curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-ripemd160-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,hmac-ripemd160,umac-128@openssh.com
```

The first Commit was just a note that we should change the Code. We could do even more to secure the Connection (Dockerfile).

**Comment by moul** [comment]
ping @scaleway/core-team

**Comment by aimxhaisse** [comment]
I'd prefer to keep this file as close as possible to the one shipped by Ubuntu ; I agree that this may be better re-ordered or without SHA-1, but that's not a decision we have to take.

**Comment by edouardb** [comment]
I agree with @aimxhaisse

**Comment by elcore** [comment]
@moul @aimxhaisse @edouardb I would suggest to reorder the configuration and update the Dokumentation as suggested by @moul, so we would have a safe connection and still support for older Clients.

**Comment by brmzkw** [comment]
I tend to agree with @aimxhaisse. Don't you think it would be preferable to open a bug report on Ubuntu side (https://bugs.launchpad.net/ubuntu/), see what they answer and follow their advices?

**Comment by elcore** [comment]
@brmzkw that´s a wise decision :-).

I reported them that. The security group for Ubuntu will be notified, so it´s not visible

**Comment by moul** [comment]
@elcore ok cool, keep us in touch

**Comment by elcore** [comment]
Hello @moul @brmzkw @aimxhaisse @edouardb,

I have a respone from the Ubuntu Team : https://bugs.launchpad.net/ubuntu/+source/openssh/+bug/1499392

"Hello Eldin, you're right that it is time to begin migrating away from SHA-1 in default OpenSSH configurations. However there is some historical baggage in parts of the launchpad infrastructure that prevented upgrading algorithms earlier. (Strictly speaking, the defaults aren't tied to launchpad but a configuration that doesn't allow developers to work out of the box is less than ideal.)

Some related bugs that might help explain the situation:
https://bugs.launchpad.net/ubuntu/+source/openssh/+bug/1445620
https://bugs.launchpad.net/ubuntu/+source/openssh/+bug/1445624
https://bugs.launchpad.net/ubuntu/+source/openssh/+bug/1445625

A site with many general guidelines that may influence more than just default keysize and hash selections: https://stribika.github.io/2015/01/04/secure-secure-shell.html

And, of course, whatever we select should be tested against Cisco gear, since there's always a bug or two with every openssh configuration change that prevents people from logging into or using Cisco equipment.

Colin, is it feasible to start making algorithm changes yet?

Thanks"

"Not yet. I'm actively working on the relevant bits of Launchpad infrastructure, and will upgrade to OpenSSH 7.1p1 after that. I *don't* intend to vary algorithm choices from upstream configuration, but 7.1 is already a fair bit better."

**Comment by moul** [comment]
Hi @elcore, nice to hear that, we are staying tuned

**Comment by elcore** [comment]
Hello @moul @brmzkw @aimxhaisse @edouardb,

##Update No. 2

"Thank you Colin, that's great news.

I think we should have a discussion about which algorithms to deprecate, when, for the whole distribution. I'd like a consistent approach to when we stop supporting md5/sha-1/rc4 etc. Of course different protocols may have different threat models so it may not be appropriate to apply a single blanket rule for any algorithm, but supporting 16.04 LTS in 2021 makes me think that we ought to be willing to cut the algorithms known to be weak today.

OpenSSH's choices for e.g. 7.1 will probably make a lot of sense for today but may make less sense in five years, when we're still supporting 7.1 but they've moved on. Other upstreams may not be as reliable as OpenSSH, either, and second guessing their choices may make more sense.

Thanks"

"Backporting algorithm tightening may make sense, but I don't want to end
 up in a situation where users are trying to deal with interoperability
 issues but none of the upstream docs make sense. If we're advocating
 specific changes that upstream aren't currently already considering, we
 should take that up with upstream."

"Hello Colin, Hello Seth,

Seth that sounds great :-). I totally agree you.

Colin and that´s the same Problem we had on Scaleway, but I am sure that we are finding a solution :-).

I would love to participate @ the discussion.

Have a nice day,

Eldin Hadzic"

**Comment by moul** [comment]
:) nice work as soon at it's merged in Ubuntu, we can bump our most recent ubuntu versions and try to apply this configuration to older images

**Comment by elcore** [comment]
@moul that sounds great :-)

**Pull request closed without merging**